cmd/kubeadm/app/util/config/cluster.go

```diff
@@ -61,12 +61,12 @@ func (ue *unretriableError) Error() string {
 }
 
 // FetchInitConfigurationFromCluster fetches configuration from a ConfigMap in the cluster
-func FetchInitConfigurationFromCluster(client clientset.Interface, w io.Writer, logPrefix string, newControlPlane, skipComponentConfigs bool) (*kubeadmapi.InitConfiguration, error) {
+func FetchInitConfigurationFromCluster(client clientset.Interface, w io.Writer, logPrefix string, criSocket string, newControlPlane, skipComponentConfigs bool) (*kubeadmapi.InitConfiguration, error) {
```

Comment:
The problem is that we are trying to do a couple of things with a single function here.
1. We pass `newControlPlane` parameter here to distinguish if we need to load just the ClusterConfiguration (+ optionally the component configs) or we need the NodeRegistrationOptions and local API address too. The fact is that the latter two are not needed in the case of `kubeadm join` and if we do `kubeadm upgrade node` on worker node and they need to be separated in another func.
2. We call `SetInitDynamicDefaults` in the func unconditionally where it should be called only if `newControlPlane == false`.

My proposal is the following:
1. Rename `getInitConfigurationFromCluster` to `FetchClusterConfigurationFromCluster`. Then change the return type to `*ClusterConfiguration`, remove the `newControlPlane` parameter from the func and move the code connected to it to the `FetchInitConfigurationFromCluster` func itself.
2. `FetchInitConfigurationFromCluster` should call `FetchClusterConfigurationFromCluster`. Then, in accordance with the `newControlPlane` parameter, it should call `getNodeRegistration`, `getAPIEndpoint`, and `SetInitDynamicDefaults`.
3. `fetchInitConfiguration` should be merged into `fetchInitConfigurationFromJoinConfiguration`. It should call `FetchClusterConfigurationFromCluster`. `fetchInitConfigurationFromJoinConfiguration` should also call `SetInitDynamicDefaults` at its end.
4. Sort of the same thing (as in 3) needs to be done for `kubeadm upgrade node` when upgrading worker node (possibly move `fetchInitConfigurationFromJoinConfiguration` right next to `FetchInitConfigurationFromCluster`?).
5. Adjust tests accordingly.

And, of course, this is caused by misusing InitConfiguration...

Reply:
Thank you for your detailed explanation @rosti . This makes sense! I'll have a try:)

Reply:
@neolit123 @rosti 
I think I've found a simple way to avoid this problem. 
Just set an irrelevant value, and then we can skip the unnecessary checks.
Maybe we can fix the problem as easily as this :)